### DISCUSSION TITLE
idea / showerthought / principle: try to standardize on C-o for invoking transient menus?

### DISCUSSION THREAD

**Discussion by dandrake** [kickingvegas/casual-suite#33] [Ideas]
I am loving the casual suite -- and the broader user interface approach with the transient menus in emacs -- and noticed that a number of them use `C-o` somehow to invoke the menu: `re-builder`, `ibuffer`, `calc`, `bookmarks`, and `info`. 

That's what I have; it's possible your documentation for those doesn't always recommend `C-o`. But I guess the point of this issue is that it should! :)

Using a consistent keybinding improves usability -- it's easier to remember, more consistent, and more discoverable.

(In fact, while writing this issue, I used that discoverability -- I didn't recall the transient/casual menu binding for `re-builder` -- so I started it and just tried `C-o`. And there it was! Hooray!)

Your casual suite advertises itself as "opinionated" so I think it's fitting to have a strong opinion that `C-o` or variations thereof be the standard, easily-discoverable way to invoke these menus.

In fact, bigger picture, as others in the emacs blogosphere (is "blogosphere" still a thing in 2024?) have mentioned, I think this overall behavior / UI / workflow should be considered the standard approach for emacs going forward. There's the venerable `C-h b` and friends to see keybindings, but these menus are a very convenient and usable middle ground between:

1. Classic emacs functionality where you just need to know all the keybindings to do stuff in your buffer, and...
2. Classic emacs help functionality: `C-h k`, `C-h w`, `C-h b`, `C-h m`, info manuals, and so on.

So in my opinion, I think these casual/transient menus should be invoked with some kind of `C-o` keybinding whenever it makes sense. I know this isn't much of an issue, but I wanted to float the idea.

**Comment by kickingvegas** [comment]
Agreed. In fact, I recently changed the instructions for Casual I-Search to use the `C-o` [binding](https://github.com/kickingvegas/casual-isearch/issues/50). Going forward the convention for Casual packages is:

-   If the main menu binding can be mode specific, use `C-o`.
    -   Auxiliary menus such as sorting and search/filtering are bound best case wrt that mode’s bindings are for now.
        -   Unclear to me at this point to unify sort binding which `s` for Dired and I-Buffer, `S` for Bookmarks.
-   If the menu binding can be globally used and overrides a popular default binding (particularly a prefix), then try to best effort duplicate the same bindings that were replaced using that prefix. This is the example with Casual Avy using the binding `M-g`.

For those few folks that really like `C-o` bound to `newline`, my sympathies but the `return` key is always there.

**Reply by jdtsmith** [reply]
`C-o` isn't newline, but `open-line`, which is different in that it preserves point.  There was a big debate on emacs-devel about `C-o` recently, and many came to its defense.  `M-o` seems pretty good though.

**Reply by dandrake** [reply]
To be clear, I'd only want `C-o` where it's not already bound to something meaningful. So in a regular text buffer (or related), I would not overwrite `C-o` for open-line; rather, I'd want it in some prefix.

For modes re-builder, ibuffer, or elsewhere, using C-o in the base mode map works well. 

And, in the end, I have no particular reason for C-o. Really I just want something consistent, easy, and discoverable. Thus far, `C-o` seems like a pretty decent candidate.

**Comment by kickingvegas** [comment]
@dandrake First let’s consider as axiomatic the ability for a user to always be able to choose their own binding to invoke a Casual menu. People are free to disregard the opinionated guidance I give. That said, here’s my take:

1.  I think for most modes, a single, common ergonomic binding to invoke a Casual menu is desired.
2.  Nearly all ergonomic bindings involving a modifier key in Emacs have been mapped to a command. It might be impossible to satisfy (1) without replacing an existing binding to a command.
3.  I’ve stated clearly with Casual that I do not feel beholden to conform to existing Emacs mappings of bindings to commands.

Given the above, I’ve found `C-o` to personally satisfy the above for me, as I rarely use `open-line`. That said, I’m open to consider other binding candidates to suggest as the universal Casual key. Among them:

`M-o`
This suggestion by @jdtsmith is one that I find compelling as it is not commonly used in Emacs. My only concern is that for macOS, M-o is commonly used as the open command. On the other hand C-o is commonly used to open documents in Windows and Linux so it isn’t a strong concern.

`F10`
While this stomps over the existing binding to invoke a mouse-driven menu, for many `F10` is unused because they rarely use the mouse menus. An objection to using function keys though would be its distance from the home row of a keyboard. 

On the subject of `open-line`, I’ve been working on a library of Transient menus designed to help out with making some infrequently and frequently used commands related to editing text. This is not yet finalized but here’s a pre-release screenshot. Consider the scenario where this menu is invoked by `C-o` and if the user wanted to run `open-line`, could do so by typing `C-o` again.

<img width="955" alt="Screenshot 2024-08-28 at 9 09 13 AM" src="https://github.com/user-attachments/assets/42c09bd1-5507-453f-8c03-f3f76984d145">
